project_euler/problem_345/sol1.py

```diff
@@ -0,0 +1,177 @@
+        if not has_booked[enhanced_matrix[i][j][1]]:
+            has_booked[enhanced_matrix[i][j][1]] = True
+
+            # Try to see the maximum value could be reached from current node.
+            # Take not booked columns, starts from (i+1)-th row
+            #   and each of them find the largest one and accumulate them.
+            # In the end of loop, combine current value with the accumulative value
+            #   and become upper bound.
+            possible_max_value = value + enhanced_matrix[i][j][0]
+            for k in range(len(has_booked)):
+                if not has_booked[k]:
+                    sub_max_value = float("-inf")
+                    for c in range(i + 1, m):
+                        if sub_max_value < matrix[c][k]:
+                            sub_max_value = matrix[c][k]
+                    possible_max_value += sub_max_value
+
+            # Stop evaluate next sub-tree
+            #   if there is no way to overcome current max value
+            if possible_max_value > max_value:
+                max_value = solve(
+                    matrix,
+                    enhanced_matrix,
+                    i + 1,
+                    has_booked,
+                    value + enhanced_matrix[i][j][0],
+                    max_value,
+                )
+
+            # Ensure current node is not booked anymore and others can use it
+            has_booked[enhanced_matrix[i][j][1]] = False
+
+    return max_value
+
+
+# Problem inputs
+matrix1 = [
+    [7, 53, 183, 439, 863],
+    [497, 383, 563, 79, 973],
+    [287, 63, 343, 169, 583],
+    [627, 343, 773, 959, 943],
+    [767, 473, 103, 699, 303],
+]
+
+matrix2 = [
+    [7, 53, 183, 439, 863, 497, 383, 563, 79, 973, 287, 63, 343, 169, 583],
+    [627, 343, 773, 959, 943, 767, 473, 103, 699, 303, 957, 703, 583, 639, 913],
+    [447, 283, 463, 29, 23, 487, 463, 993, 119, 883, 327, 493, 423, 159, 743],
+    [217, 623, 3, 399, 853, 407, 103, 983, 89, 463, 290, 516, 212, 462, 350],
+    [960, 376, 682, 962, 300, 780, 486, 502, 912, 800, 250, 346, 172, 812, 350],
+    [870, 456, 192, 162, 593, 473, 915, 45, 989, 873, 823, 965, 425, 329, 803],
+    [973, 965, 905, 919, 133, 673, 665, 235, 509, 613, 673, 815, 165, 992, 326],
+    [322, 148, 972, 962, 286, 255, 941, 541, 265, 323, 925, 281, 601, 95, 973],
+    [445, 721, 11, 525, 473, 65, 511, 164, 138, 672, 18, 428, 154, 448, 848],
+    [414, 456, 310, 312, 798, 104, 566, 520, 302, 248, 694, 976, 430, 392, 198],
+    [184, 829, 373, 181, 631, 101, 969, 613, 840, 740, 778, 458, 284, 760, 390],
+    [821, 461, 843, 513, 17, 901, 711, 993, 293, 157, 274, 94, 192, 156, 574],
+    [34, 124, 4, 878, 450, 476, 712, 914, 838, 669, 875, 299, 823, 329, 699],
+    [815, 559, 813, 459, 522, 788, 168, 586, 966, 232, 308, 833, 251, 631, 107],
+    [813, 883, 451, 509, 615, 77, 281, 613, 459, 205, 380, 274, 302, 35, 805],
+]
+
+
+def solution(matrix: list = matrix2) -> int:
```

Comment:
This function doesn't do anything except for calling the `solve` function. So just remove this function and rename `solve` -> `solution`. Read through the solution guidelines (project_euler/README.md)